compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -0,0 +1,232 @@
+    }
+    return true;
+  };
+
+  ReassociationIndices range;
+  AffineExpr preExpr;
+  for (auto nextExpr : genericOp.getIndexingMapsArray().front().getResults()) {
+    unsigned pos = nextExpr.cast<AffineDimExpr>().getPosition();
+    if (!range.empty()) {
+      if (!hasAllMapsSameSequence(preExpr, nextExpr) || !pLoops.count(pos)) {
+        if (range.size() > 1)
+          contiguousLoops.push_back({range.begin(), range.end()});
+        range.clear();
+      }
+    }
+    preExpr = nextExpr;
+    if (pLoops.count(pos)) range.push_back(pos);
+  }
+  if (range.size() > 1) contiguousLoops.push_back(range);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Collapsing dimensions if possible: ";
+    for (auto indices : contiguousLoops) {
+      llvm::dbgs() << "[";
+      for (auto idx : indices) llvm::dbgs() << idx << ",";
+      llvm::dbgs() << "]\t";
+    }
+    llvm::dbgs() << "\n";
+  });
+
+  return contiguousLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic
+static FailureOr<SmallVector<Value>> collapseLinalgGeneric(
+    IRRewriter &rewriter, linalg::GenericOp genericOp,
+    SmallVector<ReassociationIndices> &collapseIndices) {
+  rewriter.setInsertionPoint(genericOp->getParentOp());
+  FailureOr<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (failed(replacements) || replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  return replacements;
+}
+
+/// Returns true if the given op is collapsable.
+static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+  // TODO(guray) There is no mechanism to tell the collapsed indexes to
+  // `tensor.expand_shape`. Once we have this support in MLIR, we can enable
+  // dynamic tensor shapes.
+  if (genericOp.hasDynamicShape()) return false;
+
+  // TODO(guray) Currently we can only collapse when result of all the
+  // AffineMaps are dimensions. Possible to collapse cases like
+  // affine_map<d0, d1+d2> with affine_map<d0, d1+d2>, however, this is not
+  // supported in collapsing mechanism in MLIR. Once we have this support,
+  // we can remove this if statement.
+  if (llvm::any_of(genericOp.getIndexingMapsArray(), [](AffineMap map) {
+        return !map.isProjectedPermutation();
+      })) {
+    return false;
+  }
+
+  // IndexOp allows accesing induction variables. Collapsing might cause
+  // performance regression, so we disable it.
+  if (genericOp.hasIndexSemantics()) return false;
+
+  return true;
+}
+
+/// Traverses all the the Ops in DispatchRegionOps and finds linalg.generic Op
+/// without any producers.
+static FailureOr<linalg::GenericOp> findRootGenericOp(
+    DispatchRegionOp regionOp) {
+  auto &ops = regionOp.getBody().front().getOperations();
+  if (ops.size() != 2) return failure();
```

Comment:
This is too specific. I think we should just walk through the region and find the number of ops that are "compute" ops. You can treat all ops that implement `TilingInterface` as compute ops. So then you can check the number of these ops is 1?

Reply:
I implemented what you described here. On top of that I added `if (ops.size() != 2)` to make sure the root has no producer. The original code was doing similar.